File: psykit/_internal/models/events.py
from typing import Literal, Annotated, Union

import pydantic

from psykit._internal.models.fields import DatetimeUTC
from psykit._internal.models.node_engine.base import NullValue
from psykit._internal.models.node_engine.node_graph import NodeResult

from uuid import UUID

# %%
class BaseEvent(pydantic.BaseModel):
    event_id: UUID
    run_id: UUID
    event_type: str
    event_payload: NullValue
    event_timestamp: DatetimeUTC


# %%
class StartEvent(BaseEvent):
    event_type: Literal['StartEvent'] = 'StartEvent'


class EndEvent(BaseEvent):
    event_type: Literal['EndEvent'] = 'EndEvent'


# %%
class NodeResultEvent(BaseEvent):
    event_type: Literal['NodeResultEvent'] = 'NodeResultEvent'
    event_payload: NodeResult

# %%
Event = Annotated[
    Union[
        StartEvent,
        NodeResultEvent,
        EndEvent,
    ],
    pydantic.Field(discriminator='event_type')
]

File: psykit/_internal/compilers/psykit-browser/src/types/events.ts
type ISO8601 = string & { __brand: 'ISO8601' };
export type UUID = string & { __brand: 'UUID' };


// NodeResult:
export type NodeResult = {
    node_id: UUID,
    timestamp_start: ISO8601,
    timestamp_end: ISO8601,
    node_execution_index: number, // 0 for first execution, 1 for second, etc.
    action: any, // todo flow from NodePlayer
    runtime_metrics: any
}

// Base event type
export type BaseEvent<T extends string, P> = {
    event_id: UUID
    run_id: UUID
    event_type: T,
    event_payload: P,
    event_timestamp: ISO8601,
}


export type StartEvent = BaseEvent<'StartEvent', {}>
export type EndEvent = BaseEvent<'EndEvent', {}>
export type NodeResultEvent = BaseEvent<'NodeResultEvent', NodeResult>

// Union of all event types
export type Event = StartEvent | EndEvent | NodeResultEvent;


// Server responses
export type SubmitEventResponse = {
    redirect_url: string | null;
}

